subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/DependencyMetadataRules.java

```diff
@@ -18,40 +18,83 @@
 
 
 import org.gradle.api.Action;
-import org.gradle.api.artifacts.DependenciesMetadata;
-import org.gradle.api.internal.artifacts.repositories.resolver.DependenciesMetadataAdapter;
+import org.gradle.api.artifacts.DirectDependenciesMetadata;
+import org.gradle.api.artifacts.DependencyConstraintMetadata;
+import org.gradle.api.artifacts.DependencyConstraintsMetadata;
+import org.gradle.api.artifacts.DirectDependencyMetadata;
+import org.gradle.api.internal.artifacts.repositories.resolver.DirectDependenciesMetadataAdapter;
+import org.gradle.api.internal.artifacts.repositories.resolver.DependencyConstraintsMetadataAdapter;
+import org.gradle.api.specs.Spec;
 import org.gradle.internal.component.external.model.ModuleDependencyMetadata;
 import org.gradle.internal.reflect.Instantiator;
 import org.gradle.internal.typeconversion.NotationParser;
+import org.gradle.util.CollectionUtils;
 
 import java.util.ArrayList;
 import java.util.List;
 
 /**
- * A set of rules provided by the build script author (as {@link Action<DependenciesMetadata>}) that
+ * A set of rules provided by the build script author (as {@link Action< DirectDependenciesMetadata >}) that
  * are applied on the dependencies defined in variant/configuration metadata. The rules are applied
  * in the {@link #execute(List)} method when the dependencies of a variant are needed during dependency resolution.
  */
 public class DependencyMetadataRules {
-    private final Instantiator instantiator;
-    private final NotationParser<Object, org.gradle.api.artifacts.DependencyMetadata> dependencyNotationParser;
+    private static final Spec<ModuleDependencyMetadata> DEPENDENCY_FILTER = new Spec<ModuleDependencyMetadata>() {
+        @Override
+        public boolean isSatisfiedBy(ModuleDependencyMetadata dep) {
+            return !dep.isPending();
+        }
+    };
+    private static final Spec<ModuleDependencyMetadata> DEPENDENCY_CONSTRAINT_FILTER = new Spec<ModuleDependencyMetadata>() {
+        @Override
+        public boolean isSatisfiedBy(ModuleDependencyMetadata dep) {
+            return dep.isPending();
```

Comment:
:o: we seem to endorse here that Maven `optional` dependencies are going to be considered as constraints. Maybe this is what we want, maybe not. But there should be a test case covering this.

Reply:
Both `DependencyConstraintMetadataRulesTest` and `DependencyMetadataRulesIntegrationTest` test this for almost all test cases. But there was no test explicitly stating that _optionals_ are mapped to constraints in metadata rules. I added one.